### PR TITLE
ympd service: init

### LINKED ISSUE #17878: ympd service
## Issue description
I just added the ympd package (#17788), but it really should include a service, as ympd provides a web server for its UI. 

I'll work on this as soon as I can!

### PR DESCRIPTION
**Pull request by langston-barrett**
###### Motivation for this change

ympd provides a web ui, it is suitable to be run as a service.

I've installed, enabled, and tested this successfully locally, although without MPD. I've also customized different options, and they've worked.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

The service is based heavily on the mpd service.
Fixes #17878.

### CODE REVIEW

File: nixos/modules/services/audio/ympd.nix [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.ympd = {
+      after = [ "sound.target" "mpd.service" ];
+      description = "Standalone MPD Web GUI written in C";
+      wantedBy = [ "multi-user.target" ];
+      path = [ pkgs.ympd ];
+      script = "exec ympd --host ${cfg.mpd.host} --port ${toString cfg.mpd.port} --webport ${toString cfg.webPort} --user ${cfg.user}";
```
Comment (fpletz) [inline nixos/modules/services/audio/ympd.nix]: This generates a shell script that will be executed. You can use `serviceConfig.ExecStart` and drop the exec for simplification. If you use `"${pkgs.ympd}/bin/ympd"` here you can also drop the `path`.

File: nixos/modules/services/audio/ympd.nix [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    users.extraUsers = optionalAttrs (cfg.user == "ympd") (singleton {
+      name = "ympd";
+      group = cfg.group;
+      extraGroups = [ "audio" ];
```
Comment (fpletz) [inline nixos/modules/services/audio/ympd.nix]: Is the audio group really needed? Only processes that need to access the sound devices need this, but in this case only mpd would access those, right?

Also, does ympd access or save any files that need special permissions or are private? If not, a new user and group probably aren't needed and you can just run the service as nobody:nogroup.

**Comment by siddharthist** [comment]
Thanks for the feedback @fpletz! This is my first time writing a service, sorry about the obvious problems :-)

**Comment by ericsagnes** [comment]
`mkOption` definitions are missing `type` declarations, and it might also be nice for the manual to add `example` declarations.

Nitpick, as localhost is the default `mpd.host`, it might make sense to make the default `mpd.port` to `config.services.mpd.network.port` for better integration with the `mpd` service.

**Comment by siddharthist** [comment]
@ericsagnes Thanks! Done! What do you think about the `host` default? Does that also make sense?

**Comment by ericsagnes** [comment]
> Thanks! Done! What do you think about the host default? Does that also make sense?

Sure, it is consistent with usptream.

> -h, --host HOST
>   connect to mpd at host, defaults to localhost

**Comment by siddharthist** [comment]
@ericsagnes I just mean [I changed it in the most recent commit](https://github.com/NixOS/nixpkgs/pull/18371/commits/0fd1fb6b37bcdff9201a889cc19be0b560ee4a5d#diff-263e0c0b85c3cb5dc5eb45a153b15eceR31) so that the default is `config.services.mpd.network.listenAddress`, do you think that makes sense?

File: nixos/modules/services/audio/ympd.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    services.ympd = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to enable ympd, the MPD Web GUI.";
+      };
+
+      webPort = mkOption {
+        type = types.int;
```
Comment (fpletz) [inline nixos/modules/services/audio/ympd.nix]: The `--webport` of ympd can actually take much more values, including the address to listen on and can even do TLS with `"ssl://8081:/path/to/ssl.pem"`, see https://github.com/notandy/ympd#ssl-support. I think we should support this. Should we maybe change this to a string and keep the naming like upstream?

**Comment by fpletz** [comment]
Ok, this was my final nitpick, looks good! Thanks! :beers:

**Comment by siddharthist** [comment]
@fpletz Done and squashed :) thanks again for the feedback and help!

File: nixos/modules/services/audio/ympd.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.ympd = {
+      after = [ "sound.target" "mpd.service" ];
```
Comment (joachifm) [inline nixos/modules/services/audio/ympd.nix]: Note that `after` does not imply `wants`. Is that intentional?

Comment (siddharthist) [inline nixos/modules/services/audio/ympd.nix]: nope! good catch!

Comment (joachifm) [inline nixos/modules/services/audio/ympd.nix]: Hm, on second thought, when using remote mpds it makes no sense to pull in a local mpd instance ...

**Comment by siddharthist** [comment]
Removed the `after` portion as per @joachifm's suggestion!

**Pull request merged**